contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,380 @@
+
+The snapshotter checks status in VolumeSnapshot to decide whether to bind. It binds VolumeSnapshot and VolumeSnapshotData when ready.
+
+
+### Changes in PV Controller and CSI External Provisioner
+
+Both the CSI External Provisioner and the in-tree PV Controller will be modified to support provisioning a volume from a snapshot data source.
+
+`DataSource` is added in both `StorageClass` and `PersistentVolumeClaim` to represent the source of the data which is prepopulated to the provisioned volume. If `DataSource` is added to both during volume provisioning, `DataSource` in `PersistentVolumeClaim` will override `DataSource` in `StorageClass`. 
+
+The operation of the provisioning of a volume from a snapshot data source will be handled by the out-of-tree CSI External Provisioner.
+
+The in-tree PV Controller will handle the binding of the PV and PVC once they are ready.
+ 
+
+### CSI Volume Driver Snapshot Support
+
+The out-of-tree CSI Volume Driver creates a snapshot on the backend storage system or cloud provider, and calls CreateSnapshot through CSI ControllerServer and returns CreateSnapshotResponse. The out-of-tree CSI Volume Driver needs to implement the following functions:
+
+* CreateSnapshot, DeleteSnapshot, and create volume from snapshot if it supports CREATE_DELETE_SNAPSHOT.
+* ListSnapshots if it supports LIST_SNAPSHOTS.
+
+ListSnapshots can be an expensive operation because it will try to list all snapshots on the storage system. For a storage system that takes nightly periodic snapshots, the total number of snapshots on the system can be huge. Kubernetes should try to avoid this call if possible. Instead, calling ListSnapshots with a specific snapshot_id as filtering to query the status of the snapshot will be more desirable and efficient.
+
+CreateSnapshot is a synchronous function and it must be blocking until the snapshot is cut. For cloud providers that support the uploading of a snapshot as part of the creating snapshot operation, CreateSnapshot function must also be blocking until the snapshot is cut and after that it shall return an operation pending gRPC error code until the uploading process is complete.
+
+Create volume from snapshot will be handled by the CreateVolume controller function in the CSI Volume Driver.
+
+Refer to [Container Storage Interface (CSI)](https://github.com/container-storage-interface/spec) for detailed instructions on how CSI Volume Driver shall implement snapshot functions.
+
+
+## Transition to the New Snapshot Support
+
+To use the snapshot implementation in [external storage repo](https://github.com/kubernetes-incubator/external-storage/tree/master/snapshot), an external snapshot controller and external provisioner need to bedeployed.
+
+To use the new snapshot implementation for CSI, a sidecar container for the external snapshot controller needs to be deployed in addition to the sidecar container for the external provisioner.
+
```

Comment:
Maybe put the deployment differences you mentioned at the beginning into the following section "New design" "Old design" too? 

Just say something like "The following lists the main differences between this snapshot proposal and the the old snapshot support in external storage repo"?

Or we have two part, the first part talks about the user facing differences, the second part talks about deployment differences.

Reply:
Sure.